lib/irb/source_finder.rb

```diff
@@ -69,10 +69,19 @@ def initialize(irb_context)
 
     def find_source(signature, super_level = 0)
       case signature
-      when /\A(::)?[A-Z]\w*(::[A-Z]\w*)*\z/ # Const::Name
+      when /\A(::)?[A-Z]\w*(::[A-Z]\w*)*\z/ # ConstName, ::ConstName, ConstPath::ConstName
         eval_receiver_or_owner(signature) # trigger autoload
-        base = @irb_context.workspace.binding.receiver.yield_self { |r| r.is_a?(Module) ? r : Object }
-        file, line = base.const_source_location(signature)
+        *parts, name = signature.split('::', -1)
+        base = (
+          if parts.empty? # ConstName
+            find_const_owner(name)
+          elsif parts == [''] # ::ConstName
+            Object
+          else # ConstPath::ConstName
+            eval_receiver_or_owner(parts.join('::'))
```

Comment:
In this case, the result of this should be the same as `eval_receiver_or_owner(signature)`? If that's the case, maybe we can store that and use it here?

Reply:
If `signature == 'A::B::C'`, `parts.join('::')` will be `'A::B'`
We need to evaluate `A::B` in current binding and get `base = FooBar::A::B` (FooBar part depends on binding's `Module.nesting`)
Now we can find source by calling `base.const_source_location('C')`

Reply:
I see, thanks for explaining 👍